test/modules/TestResultCoordinator/Reports/TestResultFilter.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TestResultCoordinator
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Extensions.Logging;
+    using TestResultCoordinator.Reports;
+
+    /// <summary>
+    /// Longhaul tests need to filter out recent results, so recent unmatched results
+    /// that will soon be matched don't result in false failures.
+    ///
+    /// This abstraction filters out any result with a CreatedAt
+    /// within the configurable ignore threshold, whith
+    /// the exception of actual results matching a
+    /// non-filtered expected result.
+    /// </summary>
+    internal class TestResultFilter
+    {
+        ITestResultComparer<TestOperationResult> comparer;
+
+        internal TestResultFilter(ITestResultComparer<TestOperationResult> comparer)
+        {
+            this.comparer = comparer;
+        }
+
+        static readonly ILogger Logger = ModuleUtil.CreateLogger(nameof(TestResultCoordinator));
+
+        internal async Task<(IAsyncEnumerable<TestOperationResult>, IAsyncEnumerable<TestOperationResult>)> FilterResults(TimeSpan unmatchedResultTolerance, IAsyncEnumerable<TestOperationResult> expectedTestResultsEnumerable, IAsyncEnumerable<TestOperationResult> actualTestResultsEnumerable)
+        {
+            Logger.LogInformation("Filtering out recent results based on ignore threshold of {0}", unmatchedResultTolerance.ToString());
+
+            List<TestOperationResult> expectedResults = await expectedTestResultsEnumerable.ToListAsync();
+            List<TestOperationResult> actualResults = await actualTestResultsEnumerable.ToListAsync();
+            List<TestOperationResult> expectedResultsOutput = new List<TestOperationResult>();
+            List<TestOperationResult> actualResultsOutput = new List<TestOperationResult>();
+
+            DateTime startIgnoringAt = DateTime.UtcNow.Subtract(unmatchedResultTolerance);
+
+            int expectedResultCounter = 0;
+            int actualResultCounter = 0;
+            while (expectedResultCounter < expectedResults.Count && expectedResults[expectedResultCounter].CreatedAt < startIgnoringAt)
```

Comment:
There are a couple other approaches we could have taken for the filter.

*Split the expected results list based on ignore threshold, then for every expected result find its corresponding actual result if exists*
I didn't do this because we would need to have hashing or some other optimization to avoid O(n^2). Perf for the TRC is not a priority, but we potentially will have tons of results so we should avoid this if we can.

*Iterate backwards and pop elements off expected/actual lists*
The current implementation mimics the way the report generators iterate over results. These report generators assume:
- having an unexpected actual result at the very beginning is impossible
- having duplicate actual results is possible
- having trailing actual results with no matching expected results is possible

I think there are some logic issues with these assumptions (i.e. inconsistent views on when unexpected actual results are possible). But making these assumptions allows the iteration logic to be a bit simpler because we can assume that the first actual result will have an expected result match. So if the first expected and actual don't match, then we can keep over expected and the match will eventually be found.

If we iterate backwards, we can't leverage these assumptions. There is ambiguity as to whether the last actual result can be matched by iterating backwards over the expected results.